store/index.go

```diff
@@ -0,0 +1,324 @@
+
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+)
+
+// size of iterator keys stored in memory before more are fetched
+const (
+	iteratorKeyMinCacheSize = 100
+)
+
+// ErrUniqueExists is the error thrown when data is being inserted for a unique constraint value that already exists
+var (
+	indexPrefix     = ds.NewKey("_index")
+	ErrUniqueExists = errors.New("Unique constraint violation")
+	ErrNotIndexable = errors.New("Value not indexable")
+	ErrNoIndexFound = errors.New("No index found")
+)
+
+// Indexer is the interface to implement to support Model property indexes
+type Indexer interface {
+	BaseKey() ds.Key
+	Indexes() map[string]Index //[indexname]indexFunc
+}
+
+// Index is a function that returns the indexable, encoded bytes of the passed in bytes
+type Index struct {
+	IndexFunc func(name string, value []byte) (ds.Key, error)
+	Unique    bool
+}
+
+// IndexConfig stores the configuration for a given Index.
+type IndexConfig struct {
+	Path   string
+	Unique bool
+}
+
+// adds an item to the index
+func indexAdd(indexer Indexer, tx ds.Txn, key ds.Key, data []byte) error {
+	indexes := indexer.Indexes()
+	for path, index := range indexes {
+		err := indexUpdate(indexer.BaseKey(), path, index, tx, key, data, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// removes an item from the index
+// be sure to pass the data from the old record, not the new one
+func indexDelete(indexer Indexer, tx ds.Txn, key ds.Key, originalData []byte) error {
+	indexes := indexer.Indexes()
+
+	for path, index := range indexes {
+		err := indexUpdate(indexer.BaseKey(), path, index, tx, key, originalData, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// adds or removes a specific index on an item
+func indexUpdate(baseKey ds.Key, path string, index Index, tx ds.Txn, key ds.Key, value []byte,
+	delete bool) error {
+
+	valueKey, err := index.IndexFunc(path, value)
+	if valueKey.String() == "" {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	indexKey := indexPrefix.Child(baseKey).ChildString(path).Instance(valueKey.String()[1:])
+
+	data, err := tx.Get(indexKey)
+	if err != nil && err != ds.ErrNotFound {
+		return err
+	}
+
+	indexValue := make(keyList, 0)
+
+	if err != ds.ErrNotFound {
+		if index.Unique && !delete {
+			return ErrUniqueExists
+		}
+	}
+
+	if data != nil {
+		err = DefaultDecode(data, &indexValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	if delete {
+		indexValue.remove(key)
+	} else {
+		indexValue.add(key)
+	}
+
+	if len(indexValue) == 0 {
+		return tx.Delete(indexKey)
+	}
+
+	iVal, err := DefaultEncode(indexValue)
+	if err != nil {
+		return err
+	}
+
+	return tx.Put(indexKey, iVal)
+}
+
+// keyList is a slice of unique, sorted keys([]byte) such as what an index points to
+type keyList [][]byte
```

Comment:
This is simply an ordered list of keys, used to keep track of entries that match a given index value. The fact that it is ordered is only slightly optimal as used right now, but provides some nice optimizations should we wish to take advantage of this later.

Reply:
The byte encoding used here makes it pretty fast to encode and decode for use when querying. But if there is something faster than gob, we could alternatively use that (see also `store/encode.go`).